Fix merging with empty results

tmgordeeva opened the pull request:
If we try to merge responses with an empty response, we might take the RAW
format from an empty response over a format from a non-empty response.

Closes #84622

Commit:
Fix merging with empty results

If we try to merge responses with an empty response, we might take the RAW
format from an empty response over a format from a non-empty response.

Closes #84622

nik9000 commented on server/src/test/java/org/elasticsearch/action/search/SearchResponseMergerTests.java:
```diff
@@ -452,6 +455,49 @@ public void testMergeCompletionSuggestionsTieBreak() throws InterruptedException
 
+    // Test merging results where one result has a raw format, for instance if
+    // searching over multiple indexes where the field isn't mapped in all indexes.
+    public void testMergeEmptyFormat() throws InterruptedException {
```
If one field isn't mapped I think we use the mapped field as the "lead" for the merge. Does this happen in production?

not-napoleon replied:
Bit of an off the cuff read, so I could be wrong, but I'm pretty sure we don't do that.  We've had problems around this with terms over IP fields that were unmapped in some indices, where the unmapped ones came back as keywords (the default).  Often we use `value_type` to nudge it, but I don't think anything in the merge picks the right one naturally.

nik9000 replied:
It's quite possible for two aggs to have a different type and for that to have an empty type. But the "real" one will be first. Sorry for the confusion.

Commit:
Format fix

Commit:
Empty unmapped Max aggregations

Adds "unmapped" Max aggregations so empty Maxes are sorted properly when
reducing results. We can have empty Maxes with mapped fields so the method name
is a little misleading.

Commit:
REST api test for merging formats

Adds a REST test reproducing #84622. This requires new fields, so the other
tests have been restricted to the one original index in this test yaml.

nik9000 commented:
> The current version takes advantage of the logic from InternalAggregations reduce which is meant to move mapped fields to the top in merges. It does this by making empty Maxes claim to be unmapped, when they might be actually empty maxes on mapped fields, so I'm a little hesitant to take care of it this way.

That `isUnmapped` method is really only used for this so far as I can tell. @not-napoleon what would you think of renaming it?

I see other aggregations performing pretty similar routines. It also looks common to build "empty" aggregations for uncollected ordinals. My guess is we'd be better off renaming `isUnmapped` to `isEmpty` or `canLeadReduction`.

Commit:
Merge remote-tracking branch 'origin/master' into fix/merge-agg-format

Commit:
Rename leading reduction

Renames a method used to determine reduction order from isMapped because empty
results for mapped fields should also not lead reduction.

Commit:
Merge remote-tracking branch 'origin/master' into fix/merge-agg-format

Commit:
Merge remote-tracking branch 'origin/master' into fix/merge-agg-format

Commit:
Check for empty values from other nodes

tmgordeeva commented:
This is failing on a cross version check on the REST API test that's expected. I'm not sure how to disable that specifically here.

elasticsearchmachine commented:
Hi @tmgordeeva, I've created a changelog YAML for you.

Commit:
Update docs/changelog/86939.yaml

Commit:
Skip REST test on older versions

Commit:
Merge remote-tracking branch 'origin/master' into fix/merge-agg-format

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/Max.java:
```diff
@@ -22,10 +22,21 @@
+        this.mapped = true;
+    }
+
+    public static Max EmptyMax(String name, DocValueFormat formatter, Map<String, Object> metadata) {
```
could you rename this method?

nik9000 commented on server/src/test/java/org/elasticsearch/action/search/SearchResponseMergerTests.java:
```diff
@@ -452,6 +455,47 @@ public void testMergeCompletionSuggestionsTieBreak() throws InterruptedException
 
+    // Test merging results where one result has a raw format, for instance if
+    // searching over multiple indexes where the field isn't mapped in all indexes.
+    public void testMergeEmptyFormat() throws InterruptedException {
```
Maybe javadoc? It's not important but may as well be consistent.

Commit:
Merge remote-tracking branch 'origin/main' into fix/merge-agg-format

Commit:
Name/style cleanup

Commit:
Typo fix

Commit:
Merge remote-tracking branch 'origin/main' into fix/merge-agg-format
